core/src/main/scala/playground/QueryCompiler.scala

```diff
@@ -286,7 +302,76 @@ object CompilationErrorDetails {
   final case class EnumFallback(enumName: String) extends CompilationErrorDetails
 }
 
-object QueryCompiler extends SchemaVisitor[PartialCompiler] {
+object QueryCompiler {
+  val full = new TransitiveCompiler(AddDynamicRefinements) andThen QueryCompilerInternal
+
+}
+
+// Applies the underlying transformation on each node of the schema that has its own hints
+class TransitiveCompiler(underlying: Schema ~> Schema) extends (Schema ~> Schema) {
+
+  def apply[A](fa: Schema[A]): Schema[A] =
+    fa match {
+      case e @ EnumerationSchema(_, _, _, _) => underlying(e)
+      case p @ PrimitiveSchema(_, _, _)      => underlying(p)
+      case u @ UnionSchema(_, _, _, _) =>
+        underlying(u.copy(alternatives = u.alternatives.map(_.mapK(this))))
+      case BijectionSchema(s, bijection) => BijectionSchema(this(s), bijection)
+      case LazySchema(suspend)           => LazySchema(suspend.map(this.apply))
+      case RefinementSchema(underlying, refinement) =>
+        RefinementSchema(this(underlying), refinement)
+      case c @ CollectionSchema(_, _, _, _) => c.copy(member = this(c.member))
+      case m @ MapSchema(_, _, _, _) => underlying(m.copy(key = this(m.key), value = this(m.value)))
+      case s @ StructSchema(_, _, _, _) => underlying(s.copy(fields = s.fields.map(_.mapK(this))))
+    }
+
+}
+
+object AddDynamicRefinements extends (Schema ~> Schema) {
```

Comment:
**praise:** This now basically serves a similar tole to the old `DynamicSchemaIndex`'s `knownHints` - except a bit less declarative (can't think of a way to make it more like that at the moment)